File: sysMetrics.c
#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

int main(int argc, const char* argv[]){
    //Get metrics
 char freeCMD[] = "free -m | grep '+.*buffers' | awk '{print $4}'";
        char cpuCMD[100] = "grep 'cpu ' /proc/stat | awk '{print ($2+$4)*100/($2+$4+$5)}'";
        char url[] = " 'http://localhost:8086/write?db=sysMetrics'";
        char cmd[] = "curl -i -XPOST";
        int freeMem, cpuUseage;

        while (1){
        //get info
                FILE *in;
                extern FILE *popen();
                char buff[512];

                if(!(in = popen(freeCMD, "r"))){
                        exit(1);
                }

                while(fgets(buff, sizeof(buff), in)!=NULL){
                        freeMem = atoi(buff);
                }


                if(!(in = popen(cpuCMD, "r"))){
                        exit(1);
                }
                while(fgets(buff, sizeof(buff), in)!=NULL){
                        printf("CPU use: %s\n", buff);
                        cpuUseage = atoi(buff);
                }

                pclose(in);

                //CPU USEAGE
                //curl into influxDB
                char data[150];
                sprintf(data, "--data-binary 'cpuUse,host=denprdinf0017,region=denver cpuUsage=%d'", cpuUseage);
         
                char curlCMD[300];

		sprintf(curlCMD, "%s %s %s", cmd, data, url);
                system(curlCMD);

                char mdata[150];
                sprintf(mdata, "--data-binary 'freeMem,host=denprdinf0017,region=denver freeMem=%d'", freeMem);

                char memCMD[300];
                sprintf(memCMD, "%s %s %s", cmd, mdata, url);
                system(memCMD);

                sleep(1);
	}
    return 0;
}

File: sysMetrics.py
import subprocess
import time
import datetime
import re

def frMemCmd():
    p1 = subprocess.Popen(["free", "-m"], stdout=subprocess.PIPE)
    p2 = subprocess.Popen(["grep", "+*.buffers"], stdin=p1.stdout, stdout=subprocess.PIPE)
    p3 = subprocess.Popen(["awk", "{print $4}"], stdin=p2.stdout, stdout=subprocess.PIPE)

    freeResp = p3.communicate()[0]
    freeResp = freeResp.replace("\n","")

    matchObj = re.match('shared(.*)', freeResp)
    if matchObj:
        return matchObj.group(1)

def cpuUtCmd():
    p1 = subprocess.Popen(["grep", "cpu", "/proc/stat"], stdout=subprocess.PIPE)
    p2 = subprocess.Popen(["awk", "{print ($2+$4)*100/($2+$4+$5)}"], stdin = p1.stdout, stdout=subprocess.PIPE)
        #get first reading
    cpuResp = re.match('(.*)\n', p2.communicate()[0])
    return cpuResp.group()

def ioStatCmd():
        #"iostat -x | awk '{print $1,$4,$5}'"
    p1 = subprocess.Popen(["iostat", "-x"], stdout=subprocess.PIPE)
    p2 = subprocess.Popen(["awk", "{print $1,$4,$5}"], stdin=p1.stdout, stdout=subprocess.PIPE)

    ioResp = re.match("Device:\s+(\w+)", p2.communicate()[0])#\s+r/s\s+w/s.(\w+-*\d*)\s+(\d+.\d+)\s+(\d+.\d+)", p2.communicate()[0])
    print ioResp
        #return ioResp.group()

def writeCpu(perc):
    subprocess.call(["curl", "-i", "-XPOST", "http://localhost:8086/write?db=sysMetrics", "--data-binary", "cpuUse,host=denprdinf0017,region=denver cpuUse=" + perc])

def writeMem(mb):
    subprocess.call(["curl", "-i", "-XPOST", "http://localhost:8086/write?db=sysMetrics", "--data-binary", "freeMem,host=denprdinf0017,region=denver freeMem=" + mb])

if __name__ == '__main__':
    dbUrl = "'http://localhost:8086/write?db=sysMetrics'"
    
    subprocess.call(["curl", "-G", "http://localhost:8086/query", "--data-urlencode", "q=create database sysMetrics"])
    subprocess.call(["curl", "-G", "http://localhost:8086/query", "--data-urlencode", "q=create retention policy two_days on sysMetrics duration 2d Replication 1 Default"])
    
    while True:
        memInfo = frMemCmd()
        cpuInfo =  cpuUtCmd()
        writeMem(memInfo)
        writeCpu(cpuInfo)
		sleep(2)
